source/common/tcp_proxy/upstream.cc

```diff
@@ -244,31 +246,55 @@ void Http2Upstream::setRequestEncoder(Http::RequestEncoder& request_encoder, boo
   request_encoder_->getStream().addCallbacks(*this);
   const std::string& scheme =
       is_ssl ? Http::Headers::get().SchemeValues.Https : Http::Headers::get().SchemeValues.Http;
-  auto headers = Http::createHeaderMap<Http::RequestHeaderMapImpl>(
-      {{Http::Headers::get().Method, "CONNECT"},
-       {Http::Headers::get().Protocol, Http::Headers::get().ProtocolValues.Bytestream},
-       {Http::Headers::get().Scheme, scheme},
-       {Http::Headers::get().Path, "/"},
-       {Http::Headers::get().Host, hostname_}});
+  auto headers = Http::createHeaderMap<Http::RequestHeaderMapImpl>({
+      {Http::Headers::get().Method, config_.use_post() ? "POST" : "CONNECT"},
+      {Http::Headers::get().Host, config_.hostname()},
+      {Http::Headers::get().Path, "/"},
+      {Http::Headers::get().Scheme, scheme},
+  });
+
+  if (!config_.use_post()) {
+    headers->addReference(Http::Headers::get().Protocol,
+                          Http::Headers::get().ProtocolValues.Bytestream);
+  }
+
+  for (const auto& header : config_.headers()) {
+    if (header.append().value()) {
+      headers->addCopy(Http::LowerCaseString(header.header().key()), header.header().value());
+    } else {
+      headers->setCopy(Http::LowerCaseString(header.header().key()), header.header().value());
+    }
+  }
+
   const auto status = request_encoder_->encodeHeaders(*headers, false);
   // Encoding can only fail on missing required request headers.
   ASSERT(status.ok());
 }
 
 Http1Upstream::Http1Upstream(Tcp::ConnectionPool::UpstreamCallbacks& callbacks,
-                             const std::string& hostname)
-    : HttpUpstream(callbacks, hostname) {}
+                             const TunnelingConfig& config)
+    : HttpUpstream(callbacks, config) {}
 
 void Http1Upstream::setRequestEncoder(Http::RequestEncoder& request_encoder, bool) {
   request_encoder_ = &request_encoder;
   request_encoder_->getStream().addCallbacks(*this);
 
   ASSERT(request_encoder_->http1StreamEncoderOptions() != absl::nullopt);
+  ASSERT(!config_.use_post());
```

Comment:
do we have configuration validation and tests to ensure this config would be rejected?

Reply:
that sounds good, I wonder how to do config validation when the tunnel config is in the listener but the http1 options is in the cluster?

https://github.com/envoyproxy/envoy/blob/842485709d651a6057b2ffb505ffced21173e004/configs/encapsulate_in_http1_connect.yaml#L28-L38

Reply:
At a high level I think it would be easier to allow this for HTTP/1 and HTTP/2 than try to restrict HTTP/1 but I'm fine with you doing either :-)

Reply:
Done. I wonder does it violate HTTP/1 POST spec?

Reply:
Added HTTP/1 support and integration tests.